File: app/test/protocol_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:okabe_app/protocol.dart';

void main() {
  group('ServerEvent.tryParse', () {
    test('有効なイベントをパースする', () {
      final event = ServerEvent.tryParse(jsonEncode({
        'id': 42,
        'type': 'assistant_message',
        'ts': '2026-07-13T09:00:00Z',
        'payload': {'text': 'hello'},
      }));
      expect(event, isNotNull);
      expect(event!.id, 42);
      expect(event.type, 'assistant_message');
      expect(event.text, 'hello');
      expect(event.isUser, isFalse);
    });

    test('一時 error フレーム（id なし）は null', () {
      final event = ServerEvent.tryParse(jsonEncode({
        'type': 'error',
        'payload': {'message': 'invalid message'},
      }));
      expect(event, isNull);
    });

    test('JSON でない入力は null', () {
      expect(ServerEvent.tryParse('garbage'), isNull);
    });
  });

  test('encodeUserMessage はプロトコルのエンベロープを生成する', () {
    expect(jsonDecode(encodeUserMessage('やあ')), {
      'type': 'user_message',
      'payload': {'text': 'やあ'},
    });
  });
}

File: server/test/protocol.test.ts
import { describe, expect, test } from "bun:test";
import { parseClientMessage } from "../src/protocol";

describe("parseClientMessage", () => {
  test("有効な user_message を受理する", () => {
    const msg = parseClientMessage(
      JSON.stringify({ type: "user_message", payload: { text: "hello" } }),
    );
    expect(msg).toEqual({ type: "user_message", payload: { text: "hello" } });
  });

  test("JSON でない入力は null", () => {
    expect(parseClientMessage("not json")).toBeNull();
  });

  test("未知の type は null", () => {
    expect(parseClientMessage(JSON.stringify({ type: "evil", payload: { text: "x" } }))).toBeNull();
  });

  test("空文字の text は null", () => {
    expect(
      parseClientMessage(JSON.stringify({ type: "user_message", payload: { text: "" } })),
    ).toBeNull();
  });
});
